enterprise_access/apps/bffs/handlers.py

```diff
@@ -0,0 +1,396 @@
+from enterprise_access.apps.bffs.context import HandlerContext
+
+logger = logging.getLogger(__name__)
+
+
+class BaseHandler:
+    """
+    A base handler class that provides shared core functionality for different BFF handlers.
+    The `BaseHandler` includes core methods for loading data and adding errors to the context.
+    """
+
+    def __init__(self, context: HandlerContext):
+        """
+        Initializes the BaseHandler with a HandlerContext.
+        Args:
+            context (HandlerContext): The context object containing request information and data.
+        """
+        self.context = context
+
+    def load_and_process(self):
+        """
+        Loads and processes data. This method should be extended by subclasses to implement
+        specific data loading and transformation logic.
+        """
+        raise NotImplementedError("Subclasses must implement `load_and_process` method.")
+
+    def add_error(self, **kwargs):
+        """
+        Adds an error to the context.
+        Output fields determined by the ErrorSerializer
+        """
+        self.context.add_error(**kwargs)
+
+    def add_warning(self, **kwargs):
+        """
+        Adds an error to the context.
+        Output fields determined by the WarningSerializer
+        """
+        self.context.add_warning(**kwargs)
+
+
+class BaseLearnerPortalHandler(BaseHandler):
+    """
+    A base handler class for learner-focused routes.
+
+    The `BaseLearnerHandler` extends `BaseHandler` and provides shared core functionality
+    across all learner-focused page routes, such as the learner dashboard, search, and course routes.
+    """
+    def __init__(self, context):
+        """
+         Initializes the BaseLearnerPortalHandler with a HandlerContext and API clients.
+         Args:
+             context (HandlerContext): The context object containing request information and data.
+         """
+        super().__init__(context)
+        self.context = context
+        self.license_manager_client = LicenseManagerUserApiClient(self.context.request)
+
+    def load_and_process(self):
+        """
+        Loads and processes data. This is a basic implementation that can be overridden by subclasses.
+
+        The method in this class simply calls common learner logic to ensure the context is set up.
+        """
+        try:
+            # Retrieve and process subscription licenses. Handles activation and auto-apply logic.
+            # TODO: retrieve enterprise customer metadata,ENT-9629
+            # self.load_enterprise_customer()
+            self.load_and_process_subscription_licenses()
+
+            # Retrieve default enterprise courses and enroll in the redeemable ones
+            self.load_default_enterprise_courses()
+            self.enroll_in_redeemable_default_courses()
+        except Exception as e:  # pylint: disable=broad-exception-caught
+            logger.exception(f"Error loading learner portal handler: {e}")
```

Comment:
nit: I believe `logger.exception` already includes the stack trace; I don't believe including `{e}` is necessary here.